azurecompute/src/main/java/org/jclouds/azurecompute/compute/AzureComputeServiceAdapter.java

```diff
@@ -16,39 +16,275 @@
+      final String location = template.getLocation().getId();
+      final int[] inboundPorts = template.getOptions().getInboundPorts();
+
+      String storageAccountName = templateOptions.getStorageAccountName().or(generateStorageServiceName(DEFAULT_STORAGE_ACCOUNT_PREFIX));
+
+      // get or create storage service
+      StorageService storageService = tryFindExistingStorageServiceAccountOrCreate(location, storageAccountName, storageAccountType);
+      String storageServiceAccountName = storageService.serviceName();
+
+      // check existence or create virtual network
+      checkExistingVirtualNetworkNamedOrCreate(virtualNetworkName, location, subnetName, addressSpaceAddressPrefix, subnetAddressPrefix);
+
+      // add network security group to the subnet
+      if (!networkSecurityGroupNames.isEmpty()) {
+         String networkSecurityGroupName = Iterables.get(networkSecurityGroupNames, 0);
+         logger.warn("Only network security group '%s' will be applied to subnet '%s'.", networkSecurityGroupName, subnetName);
+         final NetworkSecurityGroup networkSecurityGroupAppliedToSubnet = api.getNetworkSecurityGroupApi().getNetworkSecurityGroupAppliedToSubnet(virtualNetworkName, subnetName);
+         if (networkSecurityGroupAppliedToSubnet != null) {
+            if (!networkSecurityGroupAppliedToSubnet.name().equals(networkSecurityGroupName)) {
+               logger.debug("Removing a networkSecurityGroup %s is already applied to subnet '%s' ...", networkSecurityGroupName, subnetName);
+               // remove existing nsg from subnet
+               String removeFromSubnetRequestId = api.getNetworkSecurityGroupApi().removeFromSubnet(virtualNetworkName, subnetName, networkSecurityGroupAppliedToSubnet.name());
+               String operationDescription = format("Remove existing networkSecurityGroup(%s) from subnet(%s)", networkSecurityGroupName, subnetName);
+               waitForOperationCompletion(removeFromSubnetRequestId, operationDescription);
+            }
+         }
+         // add nsg to subnet
+         logger.debug("Adding a networkSecurityGroup %s is already applied to subnet '%s' of virtual network %s ...", networkSecurityGroupName, subnetName, virtualNetworkName);
+         String addToSubnetId = api.getNetworkSecurityGroupApi().addToSubnet(virtualNetworkName, subnetName,
+                 NetworkSecurityGroup.create(networkSecurityGroupName, null, null, null));
+         String operationDescription = format("Add networkSecurityGroup(%s) to subnet(%s)", networkSecurityGroupName, subnetName);
+         waitForOperationCompletion(addToSubnetId, operationDescription);
+      }
+
+      logger.debug("Creating a cloud service with name '%s', label '%s' in location '%s'", name, name, location);
+      String createCloudServiceRequestId = api.getCloudServiceApi().createWithLabelInLocation(name, name, location);
+      waitForOperationCompletion(createCloudServiceRequestId, format("Create cloud service (%s)", name));
```

Comment:
Same here. Move all operations global to the nodes to the upper layer. Keep in mind that you might be deploying N nodes and this method will be executed concurrently for all them.